### PR TITLE
Team-level std algos: part 7

### PR DESCRIPTION
**Pull request by fnrizzi**
Co-authored-by: Cezary Skrzyński
Co-authored-by: Jakub Strzebonski

## Content 

Team-level implementation and tests for:

- `Kokkos_Move.hpp`
- `Kokkos_MoveBackward.hpp`
- `Kokkos_ShiftLeft.hpp`
- `Kokkos_ShiftRight.hpp`

## Technical comments

- the change `class` -> `typename` and the enable if defaulted in the template parameters comes from [this request](https://github.com/kokkos/kokkos/pull/6200#issuecomment-1626406816)

- the code duplication is intentional and agreed upon after the long discussion in [here](https://github.com/kokkos/kokkos/pull/6200)

- tests are run for both contiguous views and strided views

## Other comments

- all team-level PRs are tracked in: https://github.com/kokkos/kokkos/issues/6352

- this comes from breaking apart https://github.com/kokkos/kokkos/pull/5150

## Possible conflicts

Implementation-wise, this has not conflict with other team-level PRs so it could be merged independently. 
However, a trivial conflict in the CMakeLists inside the unit test occurs if another of the team-level PRs is merged before this because each team-level PR adds test using a specific "label". Resolving this is trivial. 

## Associated wiki PRs

- `Kokkos_Move.hpp`, `Kokkos_MoveBackward.hpp`: see [wiki PR 366](https://github.com/kokkos/kokkos-core-wiki/pull/393)
- `Kokkos_ShiftLeft.hpp`, `Kokkos_ShiftRight.hpp`: see [wiki PR 366](https://github.com/kokkos/kokkos-core-wiki/pull/436)

IMPORTANT: wiki PRs are not complete yet but 99% there 

The issue tracking all *wiki PRs* is [here](https://github.com/kokkos/kokkos-core-wiki/issues/379)

### CODE REVIEW

**Review by Rombur — APPROVED** [review]

**Comment by fnrizzi** [comment]
CI failure unrelated:
```
The following tests FAILED:
	 37 - Kokkos_PerformanceTest_Benchmark (Timeout)
Errors while running CTest
```

File: algorithms/src/std_algorithms/impl/Kokkos_ShiftLeft.hpp [hunk 1]
```diff
@@ -104,6 +104,40 @@ IteratorType shift_left_impl(const std::string& label, const ExecutionSpace& ex,
...
+  // execution space impl because for this team impl we are
+  // within a parallel region, so for now we solve serially
+
+  const std::size_t numElementsToMove =
+      ::Kokkos::Experimental::distance(first + n, last);
+  if (teamHandle.team_rank() == 0) {
+    for (std::size_t i = 0; i < numElementsToMove; ++i) {
+      first[i] = std::move(first[i + n]);
+    }
+  }
```
Comment (crtrott) [inline algorithms/src/std_algorithms/impl/Kokkos_ShiftLeft.hpp:135]: this needs to be a Kokkos::single

File: algorithms/src/std_algorithms/impl/Kokkos_ShiftRight.hpp [hunk 2]
```diff
@@ -104,6 +103,60 @@ IteratorType shift_right_impl(const std::string& label,
...
+  // execution space impl because for this team impl we are
+  // within a parallel region, so for now we solve serially
+
+  const std::size_t numElementsToMove =
+      ::Kokkos::Experimental::distance(first, last - n);
+  if (teamHandle.team_rank() == 0) {
+    for (std::size_t i = 0; i < numElementsToMove; ++i) {
+      last[-i - 1] = std::move(last[-n - i - 1]);
+    }
+  }
```
Comment (crtrott) [inline algorithms/src/std_algorithms/impl/Kokkos_ShiftRight.hpp:154]: needs to be a Kokkos::single

**Review by crtrott — CHANGES_REQUESTED** [review]

**Review by crtrott — APPROVED** [review]

**Comment by fnrizzi** [comment]
unrelated Openmptarget failure:
```
The following tests FAILED:
	  4 - Kokkos_CoreUnitTest_OpenMPTarget (Failed)
```	  
and hpx 
```
The following tests FAILED:
	  4 - Kokkos_CoreUnitTest_HPX (Timeout)
```

**Pull request merged**